aptos-move/block-executor/src/executor.rs

```diff
@@ -1097,6 +1126,103 @@ where
+        let num_txns = block.num_txns() as u32;
+        let executor = {
+            let _init_timer = VM_INIT_SECONDS.start_timer();
+            E::init(environment.clone(), shared_sync_params.base_view)
+        };
+
+        let _work_with_task_timer = WORK_WITH_TASK_SECONDS.start_timer();
+
+        // Shared environment used by each executor.
+        let runtime_environment = environment.runtime_environment();
+
+        let scheduler = shared_sync_params.scheduler;
+        let scheduler_wrapper = SchedulerWrapper::V2(scheduler);
+
+        loop {
+            while scheduler.commit_hooks_try_lock() {
+                // Perform sequential commit hooks.
+                while let Some((txn_idx, incarnation)) =
+                    scheduler.try_get_sequential_commit_hook()?
+                {
+                    self.prepare_and_queue_commit_ready_txn(
+                        txn_idx,
+                        incarnation,
+                        num_txns,
+                        &self.config.onchain.block_gas_limit_type,
+                        scheduler_wrapper,
+                        shared_sync_params.versioned_cache,
+                        None,
+                        shared_sync_params.last_input_output,
+                        shared_sync_params.shared_commit_state,
+                        shared_sync_params.base_view,
+                        shared_sync_params.global_module_cache,
+                        runtime_environment,
+                        start_delayed_field_id_counter,
+                        shared_sync_params.delayed_field_id_counter,
+                        &executor,
+                        block,
+                        num_workers as usize,
+                    )?;
+                }
+                scheduler.commit_hooks_unlock();
+            }
+
+            // TODO: pass worker_id to next_task.
+            match scheduler.next_task()? {
+                TaskKind::Execute(txn_idx, incarnation) => {
+                    if incarnation > num_workers.pow(2) + num_txns + 10 {
+                        // Something is wrong if we observe high incarnations (e.g. a bug
+                        // might manifest as an execution-invalidation cycle). Break out
+                        // to fallback to sequential execution.
+                        error!("Observed incarnation {} of txn {txn_idx}", incarnation);
+                        return Err(PanicOr::Or(ParallelBlockExecutionError::IncarnationTooHigh));
+                    }
+
+                    // TODO: execution logic, below in view.
+                    // #[cfg(feature = "blockstm-profiling")]
+                    // self.profiler_state.borrow().log_info(
+                    //     self.txn_idx,
+                    //     self.incarnation,
+                    //     self.worker_id,
+                    // );
+                },
+                TaskKind::PostCommitProcessing(txn_idx) => {
```

Comment:
Why this is not part of prepare_and_queue_commit_ready_txn?

Reply:
Because this can go completely parallel after the hook is completed that runs sequentially.